spring-web/src/main/java/org/springframework/http/server/reactive/DefaultServerHttpRequestBuilder.java

```diff
@@ -70,7 +71,8 @@ public DefaultServerHttpRequestBuilder(ServerHttpRequest original) {
 		Assert.notNull(original, "ServerHttpRequest is required");
 
 		this.uri = original.getURI();
-		this.headers = new HttpHeaders(original.getHeaders());
+		// original headers can be immutable, so create a copy
+		this.headers = new HttpHeaders(new LinkedMultiValueMap<>(original.getHeaders()));
```

Comment:
If the headers is copied into LinkedMultiValueMap directly, it will break the case insensitive behavior of header names.

e.g.: If the headers contain an entry `content-type: application/json`(HTTP2 requires header names with lowercase), and the result of `headers.getContentType()` or `headers.getFirst("Content-Type")` will be null forever.

Reply:
Can you raise this as a new issue?

Reply:
indeed, thanks for the feedback! I've created issue #33666 to track and fix this in RC2